apps/tai/lib/tai/markets/order_book.ex

```diff
@@ -48,150 +68,80 @@ defmodule Tai.Markets.OrderBook do
-  Return bid/asks up to the given depth. If depth is not provided it returns
-  the full order book.
-  """
-  def quotes(name, depth \\ :all), do: GenServer.call(name, {:quotes, depth: depth})
-
-  @doc """
-  Return the bid/ask at the top of the book
-  """
-  @spec inside_quote(atom, atom) :: {:ok, Quote.t()}
-  def inside_quote(venue_id, product_symbol) do
-    name = to_name(venue_id, product_symbol)
-
-    name
-    |> quotes(1)
-    |> case do
-      {:ok, book} ->
-        inside_bid = List.first(book.bids)
-        inside_ask = List.first(book.asks)
-
-        q = %Quote{
-          venue_id: venue_id,
-          product_symbol: product_symbol,
-          last_received_at: book.last_received_at,
-          last_venue_timestamp: book.last_venue_timestamp,
-          bid: inside_bid,
-          ask: inside_ask
-        }
-
-        {:ok, q}
-    end
+      %{
+        state
+        | bids: %{},
+          asks: %{},
+          last_received_at: change_set.last_received_at,
+          last_venue_timestamp: change_set.last_venue_timestamp
+      }
+      |> apply_changes(change_set.changes)
+
+    change_set.venue
+    |> Tai.Markets.ProcessQuote.to_name(change_set.symbol)
+    |> GenServer.cast({:order_book_snapshot, new_state, change_set})
+
+    {:noreply, new_state}
   end
 
-  @spec replace(t) :: :ok
-  def replace(%OrderBook{} = replacement) do
-    replacement.venue_id
-    |> OrderBook.to_name(replacement.product_symbol)
-    |> GenServer.call({:replace, replacement})
-  end
-
-  @spec update(t) :: :ok
-  def update(%OrderBook{} = changes) do
-    changes.venue_id
-    |> OrderBook.to_name(changes.product_symbol)
-    |> GenServer.call({:update, changes})
-  end
-
-  @spec to_name(atom, atom) :: atom
-  def to_name(venue_id, product_symbol), do: :"#{__MODULE__}_#{venue_id}_#{product_symbol}"
-
-  defp ordered_bids(state) do
-    state.bids
-    |> Map.keys()
-    |> Enum.sort()
-    |> Enum.reverse()
-    |> with_price_points(state.bids)
-  end
-
-  defp ordered_asks(state) do
-    state.asks
-    |> Map.keys()
-    |> Enum.sort()
-    |> with_price_points(state.asks)
-  end
-
-  defp with_price_points(prices, levels) do
-    prices
-    |> Enum.map(fn price ->
-      size = levels[price]
-      %PricePoint{price: price, size: size}
-    end)
-  end
-
-  defp take(list, :all), do: list
-
-  defp take(list, depth) do
-    list
-    |> Enum.take(depth)
-  end
-
-  defp update_side(state, side, price_levels) do
-    new_side =
-      state
-      |> Map.get(side)
-      |> Map.merge(price_levels)
-      |> Map.drop(price_levels |> drop_prices)
-
-    state
-    |> Map.put(side, new_side)
+  def handle_cast({:apply, change_set}, state) do
+    new_state =
+      %{
+        state
+        | last_received_at: change_set.last_received_at,
+          last_venue_timestamp: change_set.last_venue_timestamp
+      }
+      |> apply_changes(change_set.changes)
+
+    change_set.venue
+    |> Tai.Markets.ProcessQuote.to_name(change_set.symbol)
```

Comment:
Market quotes seems like a subset / a function of order book. Would it make sense to use `PubSub` in `ProcessQuote` and subscribe to OrderBook changes? That way `OrderBook` would not need to know about `ProcessQuote`.

Reply:
The quote is the intended high level API. If possible I want to avoid having the advisor need to know about order book changes. The quotes are required for the `markets` command`.

If you still need order book changes after multi level quotes I think it might be worth adding an `OrderBook` adapter with hooks for snapshot & apply.